x/gamm/pool-models/balancer/amm_test.go

```diff
@@ -14,12 +14,275 @@ import (
 	"github.com/osmosis-labs/osmosis/v7/x/gamm/types"
 )
 
+// calcJoinSharesTestCase defines a testcase for TestCalcSingleAssetJoin and
+// TestCalcJoinPoolShares.
+//
+// CalcJoinPoolShares calls calcSingleAssetJoin. As a result, we can reuse
+// the same test cases for unit testing both calcSingleAssetJoin and
+//  CalcJoinPoolShares with only one tokensIn.
+type calcJoinSharesTestCase struct {
+	name         string
+	swapFee      sdk.Dec
+	poolAssets   []balancer.PoolAsset
+	tokensIn     sdk.Coins
+	expectErr    bool
+	expectShares sdk.Int
+	expectLiq    sdk.Coins
+}
+
 // allowedErrRatio is the maximal multiplicative difference in either
 // direction (positive or negative) that we accept to tolerate in
 // unit tests for calcuating the number of shares to be returned by
 // joining a pool. The comparison is done between Wolfram estimates and our AMM logic.
 const allowedErrRatio = "0.0000001"
 
+// see calcJoinSharesTestCase struct definition.
+var calcSingleAssetJoinTestCases = []calcJoinSharesTestCase{
```

Comment:
Commenting to create an issue for further work.

More test cases should be added in a separate PR:
- with error return
- with another value of `tokensIn` (we consistently use 50000 which isn't the best practice